### PR TITLE
Add subject id aspect Id to to sample store results

### PR DESCRIPTION
**Pull request by annyhe**

### CODE REVIEW

**Comment by annyhe** [comment]
This pull request changes
['sadd', subjectAbsPath, aspectName] to 
['hmset',  subjectAbsPath, { subjectid: subject.id, aspects: JSON.stringify[name1, name2, ...]) }]

I considered using a Set to store aspects per subject, but a Set gets converted to an empty object on JSON.stringify:
var s = new Set('a','b')
JSON.stringify(s) -> '{}'

Feedback is welcome!

File: cache/models/samples.js [hunk 1]
```diff
@@ -466,17 +467,16 @@ function upsertOneSample(sampleQueryBodyObj, isBulk, userName) {
     // add aspect name to subject set, add sample key to sample set,
     // create/update hash of sample
     return redisClient.batch([
-      ['sadd', subjKey, aspectName],
+      ['sadd', constants.indexKey.aspect, aspectName],
```
Comment (pallavi2209) [inline cache/models/samples.js]: why this change? We add aspect to aspect index when a new aspect is created. Also, where do we add aspect to subject hash when a new sample is created?

File: cache/models/samples.js [hunk 2]
```diff
@@ -647,6 +649,7 @@ module.exports = {
       }
 
       cleanQueryBodyObj(reqBody);
+      aspectObj = aspObj;
```
Comment (pallavi2209) [inline cache/models/samples.js]: Below code instead?
`         aspectObj = sampleStore.arrayStringsToJson(
            aspObj, constants.fieldsToStringify.aspect		            aspObj, constants.fieldsToStringify.aspect
         );		         );`
Not sure why it was put inside the value check.

File: cache/models/samples.js [hunk 3]
```diff
@@ -758,23 +766,27 @@ module.exports = {
...
+      }
+
+      return redisClient.hmsetAsync(subjKey, {
+        subjectId: subject.id,
+        aspects: JSON.stringify(aspectNames),
+      });
+    })
+    .then(() => redisOps.executeBatchCmds(cmds))
+    .then(() => redisOps.getHashPromise(sampleType, sampleName))
+    .then((updatedSamp) => cleanAddAspectToSample(updatedSamp, aspectObj));
```
Comment (pallavi2209) [inline cache/models/samples.js]: Do we attach aspect in response when posting a sample in current implementation?

File: cache/models/subject.js [hunk 4]
```diff
@@ -179,7 +180,6 @@ function completeSubjectHierarchy(res, params) {
   u.setFilters(params, filters);
   return traverseHierarchy(res)
   .then(() => {
-    // once the filtering is done, reset it back.
```
Comment (pallavi2209) [inline cache/models/subject.js]: why remove comment?

File: cache/redisOps.js [hunk 5]
```diff
@@ -279,31 +279,51 @@ module.exports = {
...
 
   /**
-   * Command to check if aspect exists in subject set
+   * Promise to check if aspect exists in subject set
    * @param  {String} subjAbsPath - Subject absolute path
    * @param  {String} aspName - Aspect name
-   * @returns {Array} - Command array
+   * @returns {Boolean} - Does it exist, or not
    */
   aspExistsInSubjSetCmd(subjAbsPath, aspName) {
```
Comment (pallavi2209) [inline cache/redisOps.js]: change name to "aspExistsInSubjSet" since its not a command anymore

File: cache/models/samples.js (hunk 1 above)
Comment (annyhe) [inline cache/models/samples.js]: good point. I will take out that line.
we add aspect to subject hash with hmset instead of sadd

File: cache/models/samples.js (hunk 3 above)
Comment (annyhe) [inline cache/models/samples.js]: good catch. there isn't an aspect object attached to sample, in the current implementation of API. Will remove

File: cache/models/subject.js (hunk 4 above)
Comment (annyhe) [inline cache/models/subject.js]: no reason. Will put it back.

File: cache/redisOps.js (hunk 5 above)
Comment (annyhe) [inline cache/redisOps.js]: will do

File: cache/models/samples.js [hunk 6]
```diff
@@ -497,6 +497,7 @@ module.exports = {
    * @returns {Promise} - Resolves to a sample object
    */
   deleteSample(sampleName, userName) {
+    let cmds = [];
```
Comment (annyhe) [inline cache/models/samples.js]: will change to const

File: cache/redisOps.js [hunk 7]
```diff
@@ -279,31 +279,51 @@ module.exports = {
...
+    });
+  },
+
+  /**
+   * Promise to get aspects of a certain subject
+   * @param  {String} subjAbsPath - Subject absolute path
+   * @returns {String} - serialized aspect names of the subject
+   */
+  getAspectNamesInSubject(subjAbsPath) {
+    // TODO: refactor aspects to aspectsName,
```
Comment (annyhe) [inline cache/redisOps.js]: in next commit.

File: tests/cache/models/subjects/getHierarchy.js [hunk 8]
```diff
@@ -106,18 +106,16 @@ describe(`api: GET ${path}`, () => {
...
       .set('Authorization', token)
-      .expect(constants.httpStatus.OK)
-      .expect((res) => {
-        expect(res.body.children[0].children[0].samples).to.be
-          .an('array');
-        expect(res.body.children[0].children[0].samples).to.not
-          .empty;
-      })
-      .end((err /* , res */) => {
+      // .expect(constants.httpStatus.OK)
```
Comment (annyhe) [inline tests/cache/models/subjects/getHierarchy.js]: will remove

**Comment by annyhe** [comment]
this pull request DOES NOT include changes to the cache aspect response to make it identical to the API aspect response. A separate pull request will have the changes to make updatedAt and createdAt fields to use ISO time format, instead of the Date format.

File: cache/models/samples.js (hunk 2 above)
Comment (annyhe) [inline cache/models/samples.js]: moved the       aspectObj = sampleStore.arrayStringsToJson(
        aspObj, constants.fieldsToStringify.aspect
      );

out of the value check

**Comment by annyhe** [comment]
@iamigo @pallavi2209 amended as per comments. Please review

**Comment by annyhe** [comment]
travis CI is failing due to enzyme installation:

File: cache/models/samples.js [hunk 9]
```diff
@@ -403,7 +404,16 @@ function handleUpsertError(objectType, isBulk) {
...
+  );
   const subjAspArr = sampleName.toLowerCase().split('|');
+  const aspectName = subjAspArr[ONE];
+  const subjKey = sampleStore.toKey(
+    constants.objectType.subject, subjAspArr[ZERO]
+  );
+  let aspectObj = {};
+  let subjectId = '';
 
   if (subjAspArr.length < TWO) {
```
Comment (shriramshankar) [inline cache/models/samples.js]: Do you want to move like 407 to 414 after like 429?

File: cache/sampleStore.js [hunk 10]
```diff
@@ -125,6 +126,20 @@ function convertToISO(date) {
  * object instance or just a regular object.
  * @returns {Object} cleaned up and ready to store in redis.
  */
+function cleanSubject(subj) {
+  let retval = subj.get ? subj.get() : subj;
+  retval = removeNullsAndStringifyArrays(retval,
+    constants.fieldsToStringify.subject);
+  return retval;
+} // cleanAspect
```
Comment (iamigo) [inline cache/sampleStore.js]: // cleanSubject

File: cache/redisOps.js [hunk 11]
```diff
@@ -279,45 +298,63 @@ module.exports = {
...
+   * @returns {Promise} - to update the subject
    */
-  aspExistsInSubjSetCmd(subjAbsPath, aspName) {
-    return [
-      'sismember',
-      redisStore.toKey(subjectType, subjAbsPath),
-      aspName.toLowerCase(),
-    ];
+  addAspectNameToSubject(subjKey, subjectId, name) {
+    let isNameUnique = true;
```
Comment (shriramshankar) [inline cache/redisOps.js]: line 328 is not used.

**Comment by shriramshankar** [comment]
Looks good to me.
